rtloader/common/builtins/datadog_agent.c

```diff
@@ -392,7 +408,80 @@ static PyObject *set_check_metadata(PyObject *self, PyObject *args)
     Py_RETURN_NONE;
 }
 
-//
+/*! \fn PyObject *store_value(PyObject *self, PyObject *args)
+    \brief This function implements the `datadog_agent.store_value` method, storing
+    the value for the key.
+    \param self A PyObject* pointer to the `datadog_agent` module.
+    \param args A PyObject* pointer to a 2-ary tuple containing the key and the value to store.
+    \return A PyObject* pointer to `None`.
+
+    This function is callable as the `datadog_agent.store_value` Python method and
+    uses the `cb_store_value()` callback to retrieve the value from the agent
+    with CGO. If the callback has not been set `None` will be returned.
+*/
+static PyObject *store_value(PyObject *self, PyObject *args)
+{
+    // callback must be set
+    if (cb_store_value == NULL) {
+        Py_RETURN_NONE;
+    }
+
+    char *key, *value;
+
+    PyGILState_STATE gstate = PyGILState_Ensure();
+
+    // datadog_agent.store_value(key, value)
+    if (!PyArg_ParseTuple(args, "ss", &key, &value)) {
+      PyGILState_Release(gstate);
+      return NULL;
+    }
+
+    PyGILState_Release(gstate);
+    cb_store_value(key, value);
+
+    Py_RETURN_NONE;
+}
+
+/*! \fn PyObject *retrieve_value(PyObject *self, PyObject *args)
+    \brief This function implements the `datadog_agent.retrieve_value` method, retrieving
+    the value for the key previously stored.
+    \param self A PyObject* pointer to the `datadog_agent` module.
+    \param args A PyObject* pointer to a tuple containing the key to retrieve.
+    \return A PyObject* pointer to the value.
+
+    This function is callable as the `datadog_agent.retrieve_value` Python method and
+    uses the `cb_retrieve_value()` callback to retrieve the value from the agent
+    with CGO. If the callback has not been set `None` will be returned.
+*/
+static PyObject *retrieve_value(PyObject *self, PyObject *args)
+{
+    // callback must be set
+    if (cb_retrieve_value == NULL) {
+        Py_RETURN_NONE;
+    }
+
+    char *key;
+
+    PyGILState_STATE gstate = PyGILState_Ensure();
+
+    // datadog_agent.retrieve_value(key)
+    if (!PyArg_ParseTuple(args, "s", &key)) {
+      PyGILState_Release(gstate);
+      return NULL;
+    }
+
+    PyGILState_Release(gstate);
```

Comment:
Should we release the GIL after calling `cb_retrieve_value` and converting the result to a python string ?

Reply:
The callback happens in go land, so the GIL shouldn't be hold there, right? Most functions don't touch the GIL at all, wondering what's the proper usage.